src/Compilers/Core/Portable/Emit/EditAndContinue/DeltaMetadataWriter.cs

```diff
@@ -784,6 +812,266 @@ private void PopulateEncLogTableParameters()
+            using var enumerator = metadataReader.CustomAttributes.GetEnumerator();
+            var hasData = enumerator.MoveNext();
+
+            // We want to manually control where we are in the list of custom attributes so we only MoveNext at the end
+            while (hasData)
+            {
+                // Data from the existing CA table
+                var customAttributeHandle = enumerator.Current;
+                var parent = metadataReader.GetCustomAttribute(customAttributeHandle).Parent;
+                var parentCodedIndex = CodedIndex.HasCustomAttribute(parent);
+
+                // Data from the CA table delta we are emitting
+                var emittedParent = emittedAttributeParents[emittedIndex];
+                var emittedCodedIndex = CodedIndex.HasCustomAttribute(emittedParent);
+
+                if (parentCodedIndex > emittedCodedIndex)
+                {
+                    // If the existing data is ahead of the emitted data then it means the current emitted attribute
+                    // must be for something that wasn't in the original compilation so we'll deal with it later.
+                    skippedParents.Add(emittedParent);
+
+                    // Move our log row index on enough to skip over the remaining attributes for this parent, which will
+                    // be filled in later when we process the deltas
+                    encLogRowId += _customAttributeParentCounts[emittedParent];
+
+                    // We're skipping the emitted attribute, so move that index on, but we want to re-check the
+                    // existing attribute in case it matches our next emitted attribute
+                    emittedIndex++;
+                }
+                else if (parent == emittedParent)
+                {
+                    // If we found the attribute we want then add an EncLog table record for it in the right position
+                    // If the right position doesn't exist, then that means we found a parent for which an attribute has
+                    // been deleted
+                    if (encLogRows.Count <= encLogRowId)
+                    {
+                        // Expand our enc log table to accommodate
+                        encLogRows.Add(0);
+                        var ca = metadataReader.GetCustomAttribute(customAttributeHandle);
+                        AddDeletedCustomAttribute(ca.Parent.Kind, ca.Constructor, ca.Value);
+                    }
+                    encLogRows[encLogRowId] = rowId;
+
+                    // We matched, so decrement the count of attributes emitted for this parent
+                    _customAttributeParentCounts[emittedParent]--;
+                    // If there are no more attributes for this parent to emit, then move on to the next
+                    if (_customAttributeParentCounts[emittedParent] <= 0)
```

Comment:
><= [](http://example.com/codeflow?start=68&length=2)

Can this actually go negative?

Reply:
I don't like updating the dictionary either, but I'm not sure another approach. `emittedAttributeParents` is just the keys from the dictionary in order so wouldn't adding the counter to it just be recreating the dictionary anyway? Though I guess just recreating it as a list would at least mean we're not modifying the field. I'll have a play.

Reply:
`(EntityHandle parent, int attributeCount)[]` wouldn't work?